### PR TITLE
[ROCm] Add diskspace check for rocm CI nodes

### LINKED ISSUE #92822: ROCm machines out of disk space
### 🐛 Describe the bug

After a recent hiccup with downloading docker images, some ROCm machines have gotten into a bad state where they're out of disk space and workflows fail while pulling docker images with the error:

```
74711b60232e: Pull complete
e968de764a03: Pull complete
9aed670494cd: Download complete
failed to register layer: ApplyLayer exit status 1 stdout:  stderr: write /opt/rocm-5.3.0/share/miopen/db/gfx906_64.kdb: no space left on device
Error: Process completed with exit code 1.
```

Sample workflow:
Runner name: 'worker-rocm-amd-34'
https://github.com/pytorch/pytorch/actions/runs/3988204875/jobs/6839488621

Not sure how many other runners are affected.

### Versions

n/a

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport

### PR DESCRIPTION
**Pull request by amdfaa**
Fixes #92822

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport

### CODE REVIEW

**Review by huydhn — COMMENTED** [review]
Overall LGTM! But as this step is generic and can be used by others as well.  It makes sense to move it to an action, i.e. https://github.com/pytorch/pytorch/blob/master/.github/actions/get-workflow-job-id/action.yml, then invoke the action wherever it is needed.  The action could accept a diskspace cutoff param defaults to 70%

Also there is a minor linter check.

**Comment by huydhn** [comment]
Also please sign the CLA first

**Comment by amdfaa** [comment]
I'm signing the easycla but it's not updating on the ticket. I submitted a help request for the issue.

**Comment by huydhn** [comment]
> I'm signing the easycla but it's not updating on the ticket. I submitted a help request for the issue.

Oh once you have signed it, I would just need to invoke the check again to make it pass

**Comment by amdfaa** [comment]
Sounds good! Please check if it works.

**Comment by huydhn** [comment]
Weird, invoking `/easycla` doesn't seem to do anything.  You're right to submit a help request.  May be also confirm that: your email is the correct one used in your GitHub account.

File: .github/actions/teardown-rocm/action.yml [hunk 1]
```diff
@@ -14,13 +14,20 @@ runs:
         docker stop $(docker ps -q) || true
         # Prune all stopped containers.
         docker container prune -f
-        # Prune everything docker if there are more than 10 images (~200GB).
-        # This is easier than using a time filter, e.g., "until=24h".
-        # Might fail if a prune is already in progress by another runner.
-        image_count=$(docker images | wc -l)
-        if [[ ${image_count} -gt 10 ]]; then
-            echo "Purging all docker caches"
-            docker system prune -af || true
```
Comment (jithunnair-amd) [inline .github/actions/teardown-rocm/action.yml:23]: @jeffdaily What was the `|| true` for?

**Review by malfet — APPROVED** [review]
Looks good to me, but please make sure to sign CLA (PR authors email must be covered by CLA)

File: .github/actions/teardown-rocm/action.yml (hunk 1 above)
Comment (malfet) [reply .github/actions/teardown-rocm/action.yml:23]: To ignore errors I guess

**Comment by amdfaa** [comment]
> Weird, invoking `/easycla` doesn't seem to do anything. You're right to submit a help request. May be also confirm that: your email is the correct one used in your GitHub account.

Sounds good!

File: .github/actions/setup-rocm/action.yml [hunk 2]
```diff
@@ -57,6 +57,26 @@ runs:
...
         fi
 
+    - name: Runner health check diskspace
+      if: always()
+      shell: bash
+      run: |
+        diskspace_cutoff=70
+        diskspace=$(df -H / --output=pcent | sed -n 2p | sed 's/%//' | sed 's/ //')
+        msg="Please file an issue on pytorch/pytorch reporting the faulty runner. Include a link to the runner logs so the runner can be identified"
+        if [[ $diskspace -ge diskspace_cutoff ]] ; then
```
Comment (malfet) [inline .github/actions/setup-rocm/action.yml:67]: Nit (usually there are no need for space between condition and semicolon)
```suggestion
        if [[ $diskspace -ge diskspace_cutoff ]]; then
```

File: .github/actions/setup-rocm/action.yml [hunk 3]
```diff
@@ -57,6 +57,26 @@ runs:
...
+            docker system prune -af
+            diskspace_new=$(df -H / --output=pcent | sed -n 2p | sed 's/%//' | sed 's/ //')
+            if [[ $diskspace_new -gt diskspace_cutoff ]] ; then
```
Comment (malfet) [inline .github/actions/setup-rocm/action.yml:70]: Same as above (also, you are missing $ to dereference variable, aren't you?"
```suggestion
            if [[ "${diskspace_new}" -gt "${diskspace_cutoff}" ]]; then
```

**Review by malfet — CHANGES_REQUESTED** [review]
Hmm, actually, any tips how one can test it?
Because, afaik `$diskspace -ge diskspace_cutoff ` would always evaluate to true, aren't it?
```
$ diskspace=1 ;if [[ $diskspace -ge diskspace_cutoff ]]; then echo "Ha ha ha $diskspace"; fi
Ha ha ha 1
```

**Comment by malfet** [comment]
@ZainRizvi , Hmm, don't we have lintrunner to detect such mistakes?

**Comment by ZainRizvi** [comment]
Looking at the message on the EasyCLA error (specifically the line with the ❌, it seems like the commit had a different github email configured than your official github email.

Please see [the first link](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) in that line to see how to edit your git config so that your commit is correctly attributed to the your github account

```
✅ login: amdfaa (https://github.com/pytorch/pytorch/commit/78e3b9191ac3382a3fe2bd9dd140b979107d50ec, https://github.com/pytorch/pytorch/commit/6ae95c3815314c79dd641dec6f20c898dbd5ee7a)
❌ The commit (https://github.com/pytorch/pytorch/commit/5b14c9d5cd9bfd107c454f8108575fc9d46ccc5b). This user is missing the User's ID, preventing the EasyCLA check. [Consult GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve.For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).
```

**Comment by ZainRizvi** [comment]
> @ZainRizvi , Hmm, don't we have lintrunner to detect such mistakes?

@kit1980, do we have a shell lintrunner to catch issues like the one mentioned [here](https://github.com/pytorch/pytorch/pull/93032#pullrequestreview-1271702273)?

File: .github/actions/teardown-rocm/action.yml (hunk 1 above)
Comment (jeffdaily) [reply .github/actions/teardown-rocm/action.yml:23]: It was there to catch errors if another prune was already in progress.  Only needed when we were going to having multiple runners per host.  As it was, having two runners trying to prune each other was a race condition.  I'm glad we went back to just a single runner.  No docker command conflicts.

**Comment by amdfaa** [comment]
The support team stated the following: To resolve this [issue], the community usually make sure the GitHub Global Configs are properly configured, and run a git rebase to update the committer/author information on that specific commit.

**Comment by jithunnair-amd** [comment]
> Overall LGTM! But as this step is generic and can be used by others as well. It makes sense to move it to an action, i.e. https://github.com/pytorch/pytorch/blob/master/.github/actions/get-workflow-job-id/action.yml, then invoke the action wherever it is needed. The action could accept a diskspace cutoff param defaults to 70%
> 
> Also there is a minor linter check.

@amdfaa Can we look into this suggestion, so that we don't have the same logic repeated in two yml files?

**Comment by ZainRizvi** [comment]
> The support team stated the following: To resolve this [issue], the community usually make sure the GitHub Global Configs are properly configured, and run a git rebase to update the committer/author information on that specific commit.

Yeah, that's basically the same thing as what I referenced in [this comment](https://github.com/pytorch/pytorch/pull/93032#issuecomment-1405598822).

To more clearly see the root cause:

1. On your terminal, cd to your pytorch git repository and make sure you're in this PR's branch
2. Run `git log -5` to see the last 5 commits on the branch
3. Note the email addresses of the author on each of those commits. They prob don't match what's configured on your github account

Example from my own branch:

```
commit d055c8eae54114bc2f1d7f69c29999638b62cbce
Author: Zain Rizvi <some-email@some-domain.com> <-- This is the email to look at!
Date:   Wed Jan 25 20:11:34 2023 -0600

    Try using CMAKE_BUILD_TYPE
```

**Comment by amdfaa** [comment]
> Hmm, actually, any tips how one can test it? Because, afaik `$diskspace -ge diskspace_cutoff ` would always evaluate to true, aren't it?
> 
> ```
> $ diskspace=1 ;if [[ $diskspace -ge diskspace_cutoff ]]; then echo "Ha ha ha $diskspace"; fi
> Ha ha ha 1
> ```

I fixed this in the latest commit.

File: .github/actions/setup-rocm/action.yml [hunk 4]
```diff
@@ -56,6 +56,13 @@ runs:
             echo "$msg"
             exit 1
         fi
+    - name: Runner diskspace health check
+      uses: actions/diskspace-cleanup
+      if: always()
+      env:
+          DISKSPACE-CUTOFF: 70
+      with:
+          diskspace-cutoff: "$DISKSPACE-CUTOFF"
```
Comment (huydhn) [inline .github/actions/setup-rocm/action.yml:65]: Nit:  I wonder why not just put `diskspace-cutoff: 70` here, or better yet, use 70 as the default vallue for dispace-cleanup action.  Then you can just call

```
- name: Runner diskspace health check
  uses: actions/diskspace-cleanup
  if: always()
```

**Review by huydhn — APPROVED** [review]
LGTM!  Thank you for fixing this issue!

**Review by malfet — APPROVED** [review]

**Comment by jithunnair-amd** [comment]
@amdfaa Added ciflow/trunk label to trigger ROCm test jobs on trunk workflow. Please check them when they finish to see that the diskspace check is being executed correctly.

**Comment by amdfaa** [comment]
> Hmm, actually, any tips how one can test it?

I have locally tested this pr on jwr-20 with a lower cutoff and the script executed successfully.

**Pull request closed without merging**